### PR TITLE
 fix #1412: annotate breaches with IsResolved and recencyIndex 

### LINKED ISSUE #1412: Augment `bundleVerifiedEmails` function with logic to add properties to the `emailEntry` objects
Augment `bundleVerifiedEmails` function with logic to add properties to the `emailEntry` objects:
  - `"resolved": true|false`
  - `"recency_index": Number`

### PR DESCRIPTION
**Pull request by groovecoder**

### CODE REVIEW

File: hibp.js [hunk 1]
```diff
@@ -118,9 +118,14 @@ const HIBP = {
       if (sha1.toUpperCase() === sha1Prefix + breachedAccount.hashSuffix) {
         foundBreaches = allBreaches.filter(breach => breachedAccount.websites.includes(breach.Name));
         foundBreaches = this.filterBreaches(foundBreaches);
+
+        // NOTE: DO NOT CHANGE THIS SORT LOGIC
+        // We store breach resolutions by recency indices,
+        // so that our DB does not contain any part of any user's list of accounts
```
Comment (groovecoder) [inline hibp.js]: self-note: we should move this sorting function into a utility function and use it everywhere when we mark breaches as resolved or when we show breaches as resolved/not.

File: controllers/user.js [hunk 2]
```diff
@@ -125,16 +125,36 @@ async function add(req, res) {
   res.redirect("/user/preferences");
 }
 
-async function bundleVerifiedEmails(email, ifPrimary, id, verificationStatus, allBreaches) {
+function getResolvedBreachesForEmail(user, email) {
+  if (typeof user.resolved_breaches !== "object") {
```
Comment (lesleyjanenorton) [inline controllers/user.js]: ...and TIL you don't need parentheses!

File: hibp.js (hunk 1 above)
Comment (lesleyjanenorton) [reply hibp.js]: One thing we should probably consider is what happens if we add back in other breach types (IsSpamList, etc)?

Comment (groovecoder) [reply hibp.js]: Updated to annotate the breaches with `recencyIndex` BEFORE `filter`ing the breaches.

Comment (groovecoder) [reply hibp.js]: Great catch, BTW!

**Comment by groovecoder** [comment]
I'm going to say unlikely for now, which is a sure-fire way to make it happen ASAP. I'd like to skip that flexibility for now just to keep things as simple as possible. If we add another source we can add the complexity for resolutions then.

File: tests/controllers/user.test.js [hunk 3]
```diff
@@ -30,6 +29,71 @@ function expectResponseRenderedSubpagePartial(resp, partial) {
...
+  expect(renderCallContext.verifiedEmails.length).toEqual(2);
+  expect(renderCallContext.unverifiedEmails.length).toEqual(1);
+  expect(renderCallContext.verifiedEmails[0].breaches.length).toEqual(mockReturnedBreaches.length);
+});
+
+
+test("user getDashboard with valid user renders verified emails and resolved breaches", async () => {
+  const testSubscriberEmail = TEST_SUBSCRIBERS.firefox_account.primary_email;
+  const testSubscriber = await DB.getSubscriberByEmail(testSubscriberEmail);
+  // TODO: figure out how the F to put resolved_breaches in the test db seed?
```
Comment (groovecoder) [inline tests/controllers/user.test.js]: When I try to add a `resolved_breaches: {"firefoxaccount@test.com": [0]},` line in the `firefox_account` object of `TEST_SUBSCRIBERS` in the `db/seeds/test_subscribers.js` file, knex can't seem to insert into the `subscribers` table. 😢

**Comment by groovecoder** [comment]
Added a couple tests for `getDashboard` to at least verify the right number of breaches and are included and marked as resolved.

**Comment by groovecoder** [comment]
Closing in favor of https://github.com/mozilla/blurts-server/pull/1481.

**Pull request closed without merging**